Update testing for Python 3.14

wshanks opened the pull request:
- [ ] Closes #340
- [ ] Executed `pre-commit run --all-files` with no errors
- [ ] The change is fully covered by automated unit tests
- [ ] Documented in docs/ as appropriate
- [ ] Added an entry to the CHANGES file

wshanks committed:
Update `log` exception test to work Python 3.14

newville commented:
@wshanks Thanks.  

I think this should probably change `.github/workflows/test.yml` to drop testing 3.8 (I would vote to drop testing with 3.9 too), and add testing for 3.13.  Given the title, it probably ought to add testing for with some release candidate of 3.14.  ;).   

I do not know whether 3.14rc2 is supported with GH Actions.

wshanks committed:
Restore test for ufloat(0, 1)

wshanks committed:
Drop Python 3.8 and 3.9 from CI and add 3.13 and 3.14

wshanks committed:
Add flag to allow Python 3.14 in GitHub workflow

wshanks committed:
Set doctest to ignore exception details to avoid Python version specific content

wshanks committed:
Configure pytest to run doctests by default

wshanks commented:
Good suggestion, @newville 

Regarding dropping 3.8 and 3.9, I agree that it is a good time. 3.9 will go end of life soon and it's unlikely we would break something in 3.9 before then any way.

I was uncertain about adding 3.14 because my first pass at `pip install .[all]` locally failed because one subdependency didn't have a 3.14 wheel yet and the source build failed until I installed a build dependency but it turns out that all the GitHub runners have the build dependencies in their default configuration, so no extra set up was needed besides setting `allow-prereleases: true`.

The CI uncovered a second test failure that I hadn't seen locally because I was running `pytest` and not the full command line in the workflow that adds the doctests. The second test failure was similar to the first -- a change in wording of an exception message related to using 0 in a math function. To deal with that I added the `IGNORE_EXCEPTION_DETAIL` doctest option. Additionally, I added some options to pytest's configuration so that running just `pytest` by itself will run the doctests along with the rest of the tests.

newville reviewed (dismissed):
@wshanks Thanks -- this looks good to me.   Thanks for adding the test with 3.14 and catching the change in wording of the ZeroDivisionError.  That seems like a fragility of using `doctests` to me, but I'm glad there is a solution.   

And thanks for dropping 3.9 and 3.8.

newville commented:
@wshanks I suggest we wait a day to give others a chance to comment before merging, but this looks good to me.  Thanks.

jagerber48 commented:
I have a few comments and questions. My disclaimer is I haven't dug too deeply into everything yet so I might be missing something that has already been stated. My apologies if so.

- The PR is more about changing CI than fixing the error in the linked issue. I think the PR title should be updated to reflect that.
- I don't know if we "officially" decided the `uncertaintes` is going to support (aka test) exactly the python supported versions. This PR basically says that that is our intention. I'm almost fine with that. The one reason I hesitate is because I've found that between ~3.8 and 3.11 there are a number of convenient typing features that are nice to use but break one of these versions. In other words, if we stop testing 3.8 and 3.9 like in this PR I think there is a good chance we introduce one of these feature and break those old versions of python. I might suggest we continue to test 3.8 and 3.9 through this PR and discuss dropping support more formally elsewhere.
- I generally like how you redid the pytest config but I'm curious what `doctest_optionflags = "IGNORE_EXCEPTION_DETAIL"
` is impacting specifically that made you want to add it? I wonder if this is related to #338 and it would be more directly helped by clearing out all the warnings the tests are raising? Maybe #338 is unrelated though.
- I do not like that changing the test from `ufloat(0, 0.2)` to `ufloat(0.0, 0.2)` fixed it. `ufloat` shouldn't have different behavior depending on if the user passes in an `int` or a `float`. The behavior should be as if we cast user input to float. In my opinion we *should* cast user input to float and see if that fixes this bug rather than changing the test to make the bug go away.
- Also I'm confused how this PR addresses issue #340. I see a new test added but I don't see any diff on the test that was failing in that issue. What is going on here? Oh, looking closer I see you are initializing `exception_class` differently. Ok I do not like it again because I'm still confused. It looks like you're priming it for a `log` function with float exception, but then the `ufloat`'s are being initialized with `int`, so why does it work now?

Before getting into this I was hoping I'd be ok with this PR just going through but I have enough issues with it I'm interested to slow things down. Basically this PR looks like a band-aid fix to address #340 but (1) because of the reasons addressed in #340, I would say this is a low priority issues and (2) we are actively in the process of "cleaning up" the code and test in this package, so I would rather give a proper fix rather than a band aid that proliferates the issues we're trying to resolve.

My main recommendation would be to refactor `test_math_module()` to use `with pytest.raises...`. That will clean things up already a little bit. I also don't think there's a need to raise the exact same exception as the math module. Rather matching a regex with `pytest.raises` will suffices. Would that be sufficient to resolve these issues?

I might recommend cleaning up the test in one PR, making sure it still passes on 3.8 - 3.12 (the versions we've been supporting) and then make the version support changes and CI changes in another PR.

Sorry for putting the brakes on a little bit for this.

wshanks commented:
> The PR is more about changing CI than fixing the error in the linked issue. I think the PR title should be updated to reflect that.

The scope has grown since I first opened the PR. I updated the title to reflect that now.

> I might suggest we continue to test 3.8 and 3.9 through this PR and discuss dropping support more formally elsewhere.

That is fair. I don't feel too strongly, but the tests run quickly any way so it is not much overhead to keep testing 3.8 and 3.9. We should discuss dropping older Python versions in a separate issue. Given the tests pass, I don't see a reason not to add the new Python versions here.

> I generally like how you redid the pytest config but I'm curious what doctest_optionflags = "IGNORE_EXCEPTION_DETAIL" is impacting specifically that made you want to add it?

It is unrelated to #338. Check out the Python 3.14 logs from this commit https://github.com/lmfit/uncertainties/pull/341/commits/b2c0f2f8aa28c24de2a0f1346882a21c7d146de5 right before I set the `IGNORE_EXCPETION_DETAIL` option. In short, the Python builtin exception message changed in 3.14 so the doctest can't document the output for both Python <=3.13 and 3.14 at the same time if it compares the detailed message beyond the exception type.

> I do not like that changing the test from ufloat(0, 0.2) to ufloat(0.0, 0.2) fixed it.

I am not sure what you are referencing. I didn't make this change. I added a `math.log(0.0)` in order to capture the exception message for this call to the builtin `math` module. In my early commits, I accidentally deleted one part of the test that might have looked like that change, but I restored it.

> It looks like you're priming it for a log function with float exception, but then the ufloat's are being initialized with int, so why does it work now?

I didn't trace it exactly, but I believe `ufloat` casts its argument to `float` so it goes down the float path rather than the int path. All that has changed in Python 3.14 is the format of the exception message for `math.log` which now differs for `int` vs `float` arguments where before the message was the same for both cases. It seems fair to compare the `ufloat` case to passing a float to a `math` module function. Technically though it is a deviation in behavior regarding integer arguments.

> Before getting into this I was hoping I'd be ok with this PR just going through but I have enough issues with it I'm interested to slow things down. Basically this PR looks like a band-aid fix to address https://github.com/lmfit/uncertainties/issues/340 but (1) because of the reasons addressed in https://github.com/lmfit/uncertainties/issues/340, I would say this is a low priority issues and (2) we are actively in the process of "cleaning up" the code and test in this package, so I would rather give a proper fix rather than a band aid that proliferates the issues we're trying to resolve.

I see this PR as continuing the currently supported behavior which based on the tests is to match the exception messages. I do not see matching exception messages in the documentation though so we could deem that as not within the supported behavior that is enforced by tests if we wanted to. I agree that we should not belabor this kind of detail since there is an active refactor in progress, but I think the current change is the more minimal one for the moment.

wshanks committed:
Restore CI testing of older Python versions

jagerber48 commented:
@wshanks Skipping responding to stuff that sounds good to me.

> > I do not like that changing the test from ufloat(0, 0.2) to ufloat(0.0, 0.2) fixed it.
> 
> I am not sure what you are referencing. I didn't make this change. I added a `math.log(0.0)` in order to capture the exception message for this call to the builtin `math` module. In my early commits, I accidentally deleted one part of the test that might have looked like that change, but I restored it.

Looking at this line in `user_test.rst`.

https://github.com/lmfit/uncertainties/pull/341/files#diff-ea7984cb50c5c7559f66200d287189c9baee3cd4445449e1056ea701b35e9cf0L300

> 
> > It looks like you're priming it for a log function with float exception, but then the ufloat's are being initialized with int, so why does it work now?
> 
> I didn't trace it exactly, but I believe `ufloat` casts its argument to `float` so it goes down the float path rather than the int path. All that has changed in Python 3.14 is the format of the exception message for `math.log` which now differs for `int` vs `float` arguments where before the message was the same for both cases. It seems fair to compare the `ufloat` case to passing a float to a `math` module function. Technically though it is a deviation in behavior regarding integer arguments.

I guess my confusion here ties in with my previous point about the test change in `user_test.rst` so maybe you can take a look at the line I linked and re-address my confusion?

> > Before getting into this I was hoping I'd be ok with this PR just going through but I have enough issues with it I'm interested to slow things down. Basically this PR looks like a band-aid fix to address #340 but (1) because of the reasons addressed in #340, I would say this is a low priority issues and (2) we are actively in the process of "cleaning up" the code and test in this package, so I would rather give a proper fix rather than a band aid that proliferates the issues we're trying to resolve.
> 
> I see this PR as continuing the currently supported behavior which based on the tests is to match the exception messages. I do not see matching exception messages in the documentation though so we could deem that as not within the supported behavior that is enforced by tests if we wanted to. I agree that we should not belabor this kind of detail since there is an active refactor in progress, but I think the current change is the more minimal one for the moment.

Ok, I'm fine with the quick fix if you can help me understand the change in `user_test.rst`.

jagerber48 commented:
https://github.com/python/cpython/issues/101410 Here's the cypthon issue about changing the error message on `math.log`.

See how the `cpython` tests look for testing the new exception. I think we could try something along these lines instead in the future: https://github.com/python/cpython/pull/124299/files#diff-a39b1e44e20048f5b8207fee83df00a2541a7f02716113e6243c33165e628cd5R2506

wshanks committed:
Remove stray change to user_guide.rst

newville commented:
> I don't know if we "officially" decided the uncertaintes is going to support (aka test) exactly the python supported versions. This PR basically says that that is our intention. I'm almost fine with that. The one reason I hesitate is because I've found that between ~3.8 and 3.11 there are a number of convenient typing features that are nice to use but break one of these versions. In other words, if we stop testing 3.8 and 3.9 like in this PR I think there is a good chance we introduce one of these feature and break those old versions of python. I might suggest we continue to test 3.8 and 3.9 through this PR and discuss dropping support more formally elsewhere.

We have to be able to drop support for older Python versions.  Saying that we support "currently supported Python versions" is perfectly reasonable and a very common approach.  It means that new versions of `uncertainties` will no longer be tested on Python 3.8 or 3.9, and may fail.  But it doesn't break existing installations. 

It might be slightly overstated, but the alternative you proposed might be extrapolated to "discuss which versions are supported for every PR".   That's not exactly "madness", but it adds work.

I wish Python had a slower release cadence, but I also think that dropping testing of Python 3.8 should not be controversial. 

> I generally like how you redid the pytest config but I'm curious what doctest_optionflags = "IGNORE_EXCEPTION_DETAIL"  is impacting specifically that made you want to add it? I wonder if this is related to https://github.com/lmfit/uncertainties/issues/338 and it would be more directly helped by clearing out all the warnings the tests are raising? Maybe https://github.com/lmfit/uncertainties/issues/338 is unrelated though.

My initial thought was to drop using `doctest` ;).  I think that @wshanks found a decent workaround.

wshanks commented:
> https://github.com/python/cpython/issues/101410 Here's the cypthon issue about changing the error message on math.log.

"math.log() is a special snowflake" :slightly_smiling_face: 

> I guess my confusion here ties in with my previous point about the test change in user_test.rst so maybe you can take a look at the line I linked and re-address my confusion?

Oh, sorry, that was something I tried just out of curiosity when first seeing the doctest error and did not mean to commit. In your previous comment, I thought you were referring to a change in a `test_` module and did not think of this change because I hadn't realized I had committed it. It is not related to addressing the doctest failure. The failure is addressed just by the doctest option `IGNORE_EXCEPTION_DETAIL` that causes doctest to ignore exception messages and only compare exception types (basically it just ignores the text after the colon on exception lines). That change is kind of going the opposite way to the test module change by not checking the exact exception message but I don't think doctest is (edit) not flexible enough to support different output per Python version in a clean way.

> See how the cpython tests look for testing the new exception. I think we could try something along these lines instead in the future:

I agree that a raises context manager would be a better pattern. I think we are essentially doing the same kind of exception message test though, just with the exceptions pulled from the `math` module's equivalent exception instead of being hard-coded like the CPython case. I wonder why they use the regex version with no regex wildcards rather than a plain string match.

jagerber48 commented:
@wshanks ok great, that `user_test.rst` change was freaking me out making me think we had different behavior depending on if the input was `int` or `float`.

This PR looks good to me now after this discussion. Thanks for taking the time. Now the PR looks like (1) add support for 3.13 and 3.14 (2) minor clean up changes to CI and (3) minor change in the tests to support this special change to `math.log` exception messages.

My guess is we could probably do something better than `IGNORE_EXCEPTION_DETAIL` in the doctests but we can worry about that elsewhere. The docs are getting pretty heavily rewritten for 4.0 anyways.

newville commented:
>> https://github.com/python/cpython/issues/101410 Here's the cypthon issue about changing the error message on math.log.
> "math.log() is a special snowflake" 🙂

Thank you. LOL.

jagerber48 commented:
I think adding support for 3.13 and 3.14 deserves an entry in the changelog

wshanks committed:
Mark Python 3.13 and 3.14 as officially supported

wshanks commented:
> I think adding support for 3.13 and 3.14 deserves an entry in the changelog

Sounds good. The only changes in the PR so far had been around updating the tests to handle edge cases that came up with 3.14. Nothing was added to the code itself to support 3.13 or 3.14. Now I added an entry to the changelog along with adding the 3.13 and 3.14 trove classifiers to pyproject.toml.

By the way, in my most recent push, there was one failure [here](https://github.com/lmfit/uncertainties/actions/runs/17617077617/job/50052893572?pr=341) on the first run. I triggered a rerun so it would pass. I think I remember a similar failure recently another time but I couldn't find it in the CI log.

jagerber48 commented:
> By the way, in my most recent push, there was one failure [here](https://github.com/lmfit/uncertainties/actions/runs/17617077617/job/50052893572?pr=341) on the first run. I triggered a rerun so it would pass. I think I remember a similar failure recently another time but I couldn't find it in the CI log.

I'll open an issue where we can discuss in more detail.

edit: See #343

newville commented:
Happy to merge. Thanks @wshanks !
